contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -99,87 +133,208 @@ abstract contract BaseHarvester is Governable {
+    function _decodeUniswapV2Path(bytes calldata data, address token)
+        internal
+        view
+        returns (address[] memory path)
+    {
+        (path) = abi.decode(data, (address[]));
+
+        // Do some validation
+        require(
+            path.length >= 2 &&
+                path[0] == token &&
+                path[path.length - 1] == baseTokenAddress,
+            "Invalid Uniswap V2 path"
+        );
+    }
+
+    /**
+     * @dev Decodes the data passed into Uniswap V3 path and validates
+     *      it to make sure the path is for `token` to `baseToken`
+     *
+     * @param data Ecnoded data passed to the `setRewardTokenConfig`
+     * @param token The address of the reward token
+     * @return path The validated Uniswap V3 path
+     */
+    function _decodeUniswapV3Path(bytes calldata data, address token)
+        internal
+        view
+        returns (bytes calldata path)
+    {
+        path = data;
+
+        // Do some validation
+        require(
+            address(uint160(bytes20(data[0:20]))) == token,
+            "Invalid Reward Token in swap path"
+        );
+        require(
+            address(uint160(bytes20(data[path.length - 20:]))) ==
+                baseTokenAddress,
+            "Invalid Base Token in swap path"
+        );
+    }
+
+    /**
+     * @dev Decodes the data passed to Balancer Pool ID
+     *
+     * @param data Ecnoded data passed to the `setRewardTokenConfig`
+     * @return poolId The pool ID
+     */
+    function _decodeBalancerPoolId(bytes calldata data)
+        internal
+        pure
+        returns (bytes32 poolId)
+    {
+        (poolId) = abi.decode(data, (bytes32));
+
+        // Do some validation
+        require(poolId != bytes32(0), "Invalid Balancer Pool ID");
+    }
+
+    /**
+     * @dev Decodes the data passed to get the pool indices and
+     *      checks it against the Curve Pool to make sure it's
+     *      not misconfigured. The indices are packed into a single
+     *      uint256 for gas savings
+     *
+     * @param data Ecnoded data passed to the `setRewardTokenConfig`
+     * @param poolAddress Curve pool address
+     * @param token The address of the reward token
+     * @return indices Packed pool asset indices
+     */
+    function _decodeCurvePoolIndices(
+        bytes calldata data,
+        address poolAddress,
+        address token
+    ) internal view returns (uint256 indices) {
+        (uint256 rewardTokenIndex, uint256 baseTokenIndex) = abi.decode(
+            data,
+            (uint256, uint256)
+        );
+
+        // Pack it into one slot for gas savings
+        // Shifts left rewardTokenIndex by 128 bits and then
+        // adds baseTokenIndex to avoid overlaps
+        indices = (rewardTokenIndex << 128) + baseTokenIndex;
+
+        ICurvePool pool = ICurvePool(poolAddress);
+        require(
+            token == pool.coins(rewardTokenIndex),
+            "Invalid Reward Token Index"
+        );
+        require(
```

Comment:
same custom errors comment as above. (there are more in the file, I won't write for each to not flood with the comments too much)